crates/astria-sequencer/src/app/mod.rs

```diff
@@ -1333,3 +1195,243 @@ struct PostTransactionExecutionResult {
+            Proposal::Process(proposal_info) => proposal_info.current_tx_group = group,
+        }
+    }
+
+    fn execution_results_mut(&mut self) -> &mut Vec<ExecTxResult> {
+        match self {
+            Proposal::Prepare(proposal_info) => &mut proposal_info.execution_results,
+            Proposal::Process(proposal_info) => &mut proposal_info.execution_results,
+        }
+    }
+
+    fn execution_results(self) -> Vec<ExecTxResult> {
+        match self {
+            Proposal::Prepare(proposal_info) => proposal_info.execution_results,
+            Proposal::Process(proposal_info) => proposal_info.execution_results,
+        }
+    }
+
+    fn into_prepare(self) -> Option<PrepareProposalInformation<'a>> {
+        match self {
+            Proposal::Prepare(proposal_info) => Some(proposal_info),
+            Proposal::Process(_) => None,
+        }
+    }
+}
+
+struct PrepareProposalInformation<'a> {
+    validated_txs: Vec<bytes::Bytes>,
+    included_signed_txs: Vec<Transaction>,
+    failed_tx_count: usize,
+    execution_results: Vec<ExecTxResult>,
+    excluded_txs: usize,
+    current_tx_group: Group,
+    mempool: Mempool,
+    metrics: &'a Metrics,
+}
+
+impl<'a> PrepareProposalInformation<'a> {
+    fn new(mempool: Mempool, metrics: &'a Metrics) -> Self {
+        Self {
+            validated_txs: Vec::new(),
+            included_signed_txs: Vec::new(),
+            failed_tx_count: 0,
+            execution_results: Vec::new(),
+            excluded_txs: 0,
+            current_tx_group: Group::BundleableGeneral,
+            mempool,
+            metrics,
+        }
+    }
+}
+
+struct ProcessProposalInformation {
+    execution_results: Vec<ExecTxResult>,
+    current_tx_group: Group,
+}
+
+impl ProcessProposalInformation {
+    fn new() -> Self {
+        Self {
+            execution_results: Vec::new(),
+            current_tx_group: Group::BundleableGeneral,
+        }
+    }
+}
+
+#[instrument(skip_all)]
+async fn proposal_checks_and_tx_execution(
+    app: &mut App,
+    tx: Arc<Transaction>,
+    // `prepare_proposal_tx_execution` already has the tx hash, so we pass it in here
+    tx_hash: Option<[u8; TRANSACTION_ID_LEN]>,
+    block_size_constraints: &mut BlockSizeConstraints,
+    proposal_info: &mut Proposal<'_>,
+) -> Result<ExitContinue> {
+    let tx_group = tx.group();
+    let tx_sequence_data_bytes = tx
+        .unsigned_transaction()
+        .actions()
+        .iter()
+        .filter_map(Action::as_rollup_data_submission)
+        .fold(0usize, |acc, seq| acc.saturating_add(seq.data.len()));
+    let tx_bytes = tx.to_raw().encode_to_vec();
+    let tx_hash_base_64 =
+        telemetry::display::base64(tx_hash.unwrap_or_else(|| Sha256::digest(&tx_bytes).into()))
+            .to_string();
+    let tx_len = tx_bytes.len();
+    let debug_msg = match proposal_info {
+        Proposal::Prepare(_) => "excluding transaction",
+        Proposal::Process(_) => "transaction error",
+    };
```

Comment:
I think it's generally better to only create new variables right before they're needed rather than en masse at the start of a method.  Generally it makes the code simpler to understand, since you don't need to remember how a given variable was constructed/what it represents at the point it's used.  It also means we don't waste cycles constructing items which might never get used in the case of an early return.